miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/workflow/DefaultWorkflowManager.java

```diff
@@ -0,0 +1,146 @@
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.ProgressStep.InputType;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Workflow;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.WorkflowExecutor;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.impl.ProgressImpl;
+import uk.ac.bbsrc.tgac.miso.core.store.ProgressStore;
+import uk.ac.bbsrc.tgac.miso.service.exception.ValidationError;
+import uk.ac.bbsrc.tgac.miso.service.exception.ValidationException;
+import uk.ac.bbsrc.tgac.miso.service.security.AuthorizationManager;
+import uk.ac.bbsrc.tgac.miso.service.workflow.factory.BarcodableProgressStepFactory;
+import uk.ac.bbsrc.tgac.miso.service.workflow.factory.IntegerProgressStepFactory;
+import uk.ac.bbsrc.tgac.miso.service.workflow.factory.ProgressStepFactory;
+
+@Service
+@Transactional(rollbackFor = Exception.class)
+public class DefaultWorkflowManager implements WorkflowManager {
+  @Autowired
+  private AuthorizationManager authorizationManager;
+
+  @Autowired
+  private ProgressStore progressStore;
+
+  @Autowired
+  private WorkflowExecutor workflowExecutor;
+
+  @Autowired
+  private BarcodableProgressStepFactory barcodableProgressStepFactory;
+
+  @Autowired
+  private IntegerProgressStepFactory integerProgressStepFactory;
+
+  @Override
+  public Workflow beginWorkflow(String workflowNameString) throws IOException {
+    WorkflowName workflowName = WorkflowName.valueOf(workflowNameString);
+
+    Progress progress = new ProgressImpl();
+    progress.setWorkflowName(workflowName);
+
+    saveProgress(progress);
+    return workflowName.createWorklow(progress);
+  }
+
+  @Override
+  public Workflow processInput(Workflow workflow, String input) throws IOException {
+    workflow.processInput(makeProgressStep(input, workflow.getNextStep().getDataTypes()));
+    saveProgress(workflow.getProgress());
+    return workflow;
+  }
+
+  @Override
+  public Workflow processInput(Workflow workflow, int stepNumber, String input) throws IOException {
+    workflow.processInput(stepNumber, makeProgressStep(input, workflow.getNextStep().getDataTypes()));
+    saveProgress(workflow.getProgress());
+    return workflow;
+  }
+
+  @VisibleForTesting
+  protected ProgressStep makeProgressStep(String input, Set<InputType> inputTypes) throws IOException {
+    for (FactoryType factoryType : getFactoryTypes(inputTypes)) {
+      ProgressStep step = makeFactory(factoryType).create(input, inputTypes);
+      if (step != null) {
+        return step;
+      }
+    }
+
+    throw new ValidationException(Collections.singletonList(new ValidationError("Could not construct ProgressStep")));
+  }
+
+  /**
+   * @return sorted, de-duplicated list of FactoryTypes
+   */
+  private List<FactoryType> getFactoryTypes(Set<InputType> inputTypes) {
+    return inputTypes.stream().map(InputType::getFactoryType).distinct().sorted().collect(Collectors.toList());
+  }
+
+  private ProgressStepFactory makeFactory(FactoryType factoryType) {
+    switch (factoryType) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this "switch" statement by "if" statements to increase readability. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1301)